ocw/lib/gce.py

```diff
@@ -3,31 +3,39 @@
 import googleapiclient.discovery
 from google.oauth2 import service_account
 from dateutil.parser import parse
+from .decorators import filterService
+from googleapiclient.errors import HttpError
 import re
 
 
 class GCE(Provider):
     __instances = dict()
 
-    def __new__(cls, vault_namespace):
-        if vault_namespace not in GCE.__instances:
-            GCE.__instances[vault_namespace] = self = object.__new__(cls)
-            self.__credentials = GCECredential(vault_namespace)
-            self.__compute_client = None
-            self.__project = None
+    def __new__(cls, vault_namespace, *args, **kwargs):
+        if vault_namespace in GCE.__instances:
+            return GCE.__instances[vault_namespace]
+        GCE.__instances[vault_namespace] = object.__new__(cls)
         return GCE.__instances[vault_namespace]
 
+    def __init__(self, namespace):
+        self.__credentials = GCECredential(namespace)
+        self.__compute_client = None
+        self.__project = None
+        self.credentials = None
+        super().__init__(namespace)
+
     def compute_client(self):
         if self.__credentials.isExpired():
             self.__credentials.renew()
             self.__compute_client = None
+            self.credentials = None
         self.__project = self.__credentials.getPrivateKeyData()["project_id"]
         if self.__compute_client is None:
-            credentials = service_account.Credentials.from_service_account_info(
+            self.credentials = service_account.Credentials.from_service_account_info(
```

Comment:
```
def iam_client(self):
        if self.__credentials.isExpired():
            self.__credentials.renew()
            self.__compute_client = None
        self.__project = self.__credentials.getPrivateKeyData()["project_id"]
        if self.__iam_client is None:
            credentials = service_account.Credentials.from_service_account_info(self.__credentials.getPrivateKeyData()
            )
            self.__iam_client = googleapiclient.discovery.build(
                "iam", "v1", credentials=credentials, cache_discovery=False
            )
        return self.__iam_client
```

this is expected version . Your approach  will lead to some problems which we already had in PCW and which cause a lot issues with stalled credentials and use  already expired credentials. 

BTW `iam_client` and `compute_client` are good candidates for use of your favorite decorator pattern . this can be done as next step

Reply:
*this change is ready for review*